cmd/jujud/updateseries/updateseries.go

```diff
@@ -118,38 +124,96 @@ func (c *UpdateSeriesCommand) Run(ctx *cmd.Context) error {
 		return err
 	}
 
+	jujuVersion, err := agent.GetJujuVersion(c.machineAgent, c.dataDir)
+	if err != nil {
+		return err
+	}
+
 	switch {
 	case toInitSys == service.InitSystemUpstart:
 		return errors.NewNotSupported(nil, "downgrade to series using upstart not supported")
 	case toInitSys == fromInitSys && toInitSys == service.InitSystemSystemd:
-		if err = c.copyTools(); err != nil {
+		if err := c.manager.CopyAgentBinary(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries, c.fromSeries, jujuVersion); err != nil {
 			return err
 		} else {
-			ctx.Infof("successfully copied tools and relinked agent tools")
+			ctx.Infof("successfully copied binaries and relinked agent binaries")
 		}
 		if !c.startAgents {
 			break
 		}
-		if err = c.startAllAgents(ctx); err != nil {
+
+		startedMachineName, startedUnitNames, err := c.manager.StartAllAgents(
+			c.machineAgent,
+			c.unitAgents,
+			c.dataDir,
+			c.toSeries,
+		)
+		if err != nil {
 			return err
+		} else {
+			for _, unitName := range startedUnitNames {
+				ctx.Infof("started %s service", unitName)
+			}
+			ctx.Infof("started %s service", startedMachineName)
 		}
 		ctx.Infof("all agents successfully restarted")
 	case toInitSys == service.InitSystemSystemd:
 		errorHappened := false
-		if err = c.writeSystemdAgents(ctx); err != nil {
+		failedAgentNames = failedAgentNames[:0]
+		startedSysdServiceNames, startedSymServiceNames, failedAgentNames, err := c.manager.WriteSystemdAgents(
+			c.machineAgent,
+			c.unitAgents,
+			c.dataDir,
+			systemdDir,
+			systemdMultiUserDir,
+			c.toSeries,
+		)
+		if err != nil {
+			for _, agentName := range failedAgentNames {
+				ctx.Warningf("failed to write service for %s: %s", agentName, err)
+			}
 			ctx.Warningf("%s", err)
 			errorHappened = true
+		} else {
+			for _, sysSvcName := range startedSysdServiceNames {
+				ctx.Infof("wrote %s agent, enabled and linked by systemd", sysSvcName)
+				for _, symSvcName := range startedSymServiceNames {
```

Comment:
i'm not sure why there is a nested loop here, esp. given the redundant  loop at line 185.  suggest trying a version of TestRunTwiceFailFirstWriteService without the line:
"s.manager = service.NewSystemdServiceManager(func() (bool, error) { return false, nil })" and symlink replaced with systemd.